pkg/globalnet/controllers/ipam/global_egressip_controller.go

```diff
@@ -0,0 +1,174 @@
+	submarinerv1 "github.com/submariner-io/submariner/pkg/apis/submariner.io/v1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/equality"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/klog"
+)
+
+func NewGlobalEgressIPController(config syncer.ResourceSyncerConfig, pool *IPPool) (ResourceController, error) {
+	var err error
+
+	klog.Info("Creating GlobalEgressIP controller")
+
+	// TODO - get list of existing GlobalEgressIPs and prime the IPPool cache.
+
+	controller := &globalEgressIPController{pool: pool}
+	controller.resourceSyncer, err = syncer.NewResourceSyncer(&syncer.ResourceSyncerConfig{
+		Name:                "GlobalEgressIP syncer",
+		ResourceType:        &submarinerv1.GlobalEgressIP{},
+		SourceClient:        config.SourceClient,
+		SourceNamespace:     corev1.NamespaceAll,
+		RestMapper:          config.RestMapper,
+		Federator:           federate.NewUpdateStatusFederator(config.SourceClient, config.RestMapper, corev1.NamespaceAll),
+		Scheme:              config.Scheme,
+		Transform:           controller.process,
+		ResourcesEquivalent: areSpecsEquivalent,
+	})
+
+	if err != nil {
+		return nil, err
+	}
+
+	return controller, nil
+}
+
+func (c *globalEgressIPController) Start(stopCh <-chan struct{}) error {
+	klog.Info("Starting GlobalEgressIP controller")
+
+	return c.resourceSyncer.Start(stopCh)
+}
+
+func (c *globalEgressIPController) process(from runtime.Object, numRequeues int, op syncer.Operation) (runtime.Object, bool) {
+	globalEgressIP := from.(*submarinerv1.GlobalEgressIP)
+
+	klog.Infof("Processing %sd %#v", op, globalEgressIP)
+
+	switch op {
+	case syncer.Create:
+		prevStatus := globalEgressIP.Status
+		requeue := c.onCreate(globalEgressIP)
+
+		return checkGlobalEgressIPStatusChanged(&prevStatus, &globalEgressIP.Status, globalEgressIP), requeue
+	case syncer.Update:
+		// TODO handle update
+	case syncer.Delete:
+		// TODO handle delete
+		return nil, false
+	}
+
+	return nil, false
+}
+
+func (c *globalEgressIPController) onCreate(globalEgressIP *submarinerv1.GlobalEgressIP) bool {
+	key, _ := cache.MetaNamespaceKeyFunc(globalEgressIP)
+
+	requeue := allocateIPs(key, globalEgressIP.Spec.NumberOfIPs, c.pool, &globalEgressIP.Status)
+	if requeue {
+		return requeue
+	}
+
+	// TODO - add IP table rules for the allocated IPs?
+
+	// TODO - start pod watcher?
+
+	return false
+}
+
+func allocateIPs(key string, numberOfIPs *int, pool *IPPool, status *submarinerv1.GlobalEgressIPStatus) bool {
+	if numberOfIPs == nil {
+		one := 1
+		numberOfIPs = &one
+	}
+
+	if *numberOfIPs == len(status.AllocatedIPs) {
+		return false
+	}
+
+	// TODO - remove IP table rules for previous allocated IPs?
+
+	klog.Infof("Allocating %d global IP(s) for %q", *numberOfIPs, key)
+
+	status.AllocatedIPs = make([]string, 0, *numberOfIPs)
+
+	for i := 0; i < *numberOfIPs; i++ {
```

Comment:
Add a `TODO` here to use the new `AllocateIPBlock` method when it is merged. `Allocate` will not ensure contiguous allocation.

Reply:
yeah. I have to review your PR.